test/test.go

```diff
@@ -0,0 +1,519 @@
+		Request:         request.Request{b},
+		Response:        response.Response{b},
+		Router:          router.Router{b},
+		IP:              ip.Ip{b},
+		Node:            node.Node{b},
+		Service:         service.Service{b},
+		ServiceRequest:  service_request.Request{b},
+		ServiceResponse: service_response.Response{b},
+	}
+	return
+}
+
+func mergeHeaders(h http.Header, in map[string][]string) {
+	for k, l := range in {
+		h.Del(k)
+		for _, v := range l {
+			h.Add(k, v)
+		}
+	}
+}
+
+func (e testEnv) noErr(err error) {
+	if err != nil {
+		e.t.Error(err)
+	}
+}
+
+func (e testEnv) Errorf(format string, args ...interface{}) {
+	e.t.Errorf(format, args...)
+}
+
+func (e *testEnv) Handle(method string, args_d []byte) []byte {
+	var out proto.Message
+	var err error
+
+	switch method {
+
+	case "kong.client.get_ip", "kong.client.get_forwarded_ip":
+		out = bridge.WrapString("10.10.10.1")
+
+	case "kong.client.get_port", "kong.client.get_forwarded_port":
+		out = &kong_plugin_protocol.Int{V: 443}
+
+	case "kong.client.get_credential":
+		out = &kong_plugin_protocol.AuthenticatedCredential{Id: "000:00", ConsumerId: "000:01"}
+
+	case "kong.client.load_consumer", "kong.client.get_consumer":
+		out = &kong_plugin_protocol.Consumer{Id: "001", Username: "Jon Doe"}
+
+	case "kong.client.authenticate":
+		// accept anything
+
+	case "kong.client.get_protocol":
+		out = bridge.WrapString("https")
+
+	case "kong.ip.is_trusted":
+		out = &kong_plugin_protocol.Bool{V: true}
+
+	case "kong.log.alert", "kong.log.crit", "kong.log.err", "kong.log.warn",
+		"kong.log.notice", "kong.log.info", "kong.log.debug":
+		args := new(structpb.ListValue)
+		e.noErr(proto.Unmarshal(args_d, args))
+		e.t.Logf("Log (%s): %v", method[strings.LastIndex(method, ".")+1:], args.AsSlice())
+
+	case "kong.node.get_id":
+		out = bridge.WrapString("a9777ac2-57e6-482b-a3c4-ef3d6ca41a1f")
+
+	case "kong.node.get_memory_stats":
+		out = &kong_plugin_protocol.MemoryStats{
+			LuaSharedDicts: &kong_plugin_protocol.MemoryStats_LuaSharedDicts{
+				Kong: &kong_plugin_protocol.MemoryStats_LuaSharedDicts_DictStats{
+					AllocatedSlabs: 1027,
+					Capacity:       4423543,
+				},
+				KongDbCache: &kong_plugin_protocol.MemoryStats_LuaSharedDicts_DictStats{
+					AllocatedSlabs: 4093,
+					Capacity:       3424875,
+				},
+			},
+			WorkersLuaVms: []*kong_plugin_protocol.MemoryStats_WorkerLuaVm{
+				{HttpAllocatedGc: 123456, Pid: 543},
+				{HttpAllocatedGc: 345678, Pid: 876},
+			},
+		}
+
+	case "kong.request.get_scheme":
+		u, err := url.Parse(e.ClientReq.Url)
+		e.noErr(err)
+		out = bridge.WrapString(u.Scheme)
+
+	case "kong.request.get_host":
+		u, err := url.Parse(e.ClientReq.Url)
+		e.noErr(err)
+		out = bridge.WrapString(u.Hostname())
+
+	case "kong.request.get_port":
+		u, err := url.Parse(e.ClientReq.Url)
+		e.noErr(err)
+		out = &kong_plugin_protocol.Int{V: getPort(u)}
+
+	case "kong.request.get_forwarded_scheme":
+		u, err := e.ClientReq.getForwardedUrl()
```

Comment:
should this be the raw value of `X-Forwarded-Proto` instead (and the following two being
`X-forwarded-host` and `X-forwarded-port`)

Reply:
done!